File: 23_sep/1584. Min Cost to Connect All Points/1584. Min Cost to Connect All Points.cpp
class Solution {
public:
    int par[1005], sz[1005];
    int find(int u){
        return par[u] = (par[u] == u ? u : find(par[u]));
    }

    bool join(int u, int v){
        u = find(u), v=find(v);
        if(u==v) return false;

        if (sz[u]<sz[v]) swap(u, v);

        par[v] = u;
        sz[u] += sz[v];
        return true;
    }
    int minCostConnectPoints(vector<vector<int>>& points) {
        int n = points.size();
        vector<pair<int, pair<int, int>>> edges;
        for(int i=0;i<n;i++){
            for(int j=i+1;j<n;j++){
                int dist = abs(points[i][0] - points[j][0]) + abs(points[i][1] - points[j][1]);
                edges.push_back({dist, {i, j}});
            }
        }
        sort(edges.begin(), edges.end());
        iota(par, par+n, 0);
        fill(sz, sz+n, 1);
        int ed=0, ans=0;
        for (auto e : edges){
            int dist = e.first, u=e.second.first, v=e.second.second;
            if (join(u, v)){
                ans+=dist;
                ed++;
            }
            if(ed==(n-1)) return ans;
        }
        return 0;
    }
};

File: 23_sep/1584. Min Cost to Connect All Points/1584. Min Cost to Connect All Points.py
class DSU: 
    def __init__(self, n): 
        self.par = [*range(n)]
        self.sz = [1] * n
    
    def find (self, u): 
        if self.par[u]!=u: 
            self.par[u] = self.find(self.par[u])
        return self.par[u]
    
    def join(self, u, v): 
        u , v = self.find(u), self.find(v)

        if(u==v): 
            return False
        if self.sz[v]>self.sz[u]: 
            v, u = u, v
        
        self.par[v] = u
        self.sz[u] += self.sz[v]
        return True
        
class Solution:
    
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        edges = []

        for i in range(n): 
            for j in range(i+1, n):
                d = abs(points[i][0]-points[j][0]) + abs(points[i][1]-points[j][1])
                edges.append([d, i, j])

        edges = sorted(edges)

        n_edges, ans = 0, 0
        dsu = DSU(n)
        for e in edges:
            if dsu.join(e[1], e[2]): 
                ans += e[0]
                n_edges+=1

            if n_edges==(n-1): 
                return ans

        return 0

